[2단계 - 자동차 경주 구현] 돌범(조재성) 미션 제출합니다.

is2js opened the pull request:
안녕하세요 코니! 
1차에서 고민해보았던 DTO를 도입하고, name과 position을 VO로 도입하면서 MVC구조를 잡아보았습니다!

### 중복 검증과 생성자에 대한 고민

1. private생성자를 기본적으로 정의하고
2. inputView에서 들어온 값(기본입력 검증만 거친)에 대해서는 
    1. 정적팩토리메소의 네이밍을  `fromInput` + 
    2. `자체검증`만 (도메인 객체내 자체검증)
    
3. 테스트 코드 등 프로덕션 코드에서 생성할 경우 
    1. 정적팩토리메소드를  `from` 으로 네이밍 +
    2. `자체검증` + 
    3. inputView에서 해주던 유틸화된 `기본 검증` 추가

이런식으로 제 나름대로 고민하고 법칙을 세웠습니다.  
그리고 `예외발생 시 재입력`하기 위한 `재귀+ try/catch`하는 부분을  **inputView**의 `getInput`만 감싸던 것에서 -> **main메서드** `getInput + 객체생성`까지 확장하여 감싸도록 변형하여
- inputView의 getInput에서도메인관련 검증까지 모조리 하던 구조 -> 도메인 객체 자체 검증이 필요한 부분은 생성자에서 처리되도록 **검증의 종류를 `콘솔input용검증ex>null, empty` + `도메인자체검증ex> 이름은 1~5자이상`  나누어 분리하는 식으로 설계**하였습니다.

### DTO, VO도입 후 리팩토링에 따른 새로운 질문들
1. Dto는 도메인보다는 Controller에 가깝다고 공부하였습니다. 그러나, Dto를 사용하는 곳은 RacingGame이며 이놈이 의존성을 가지고 있다면, **CarDto도 `domain`패키지에 넣어야할까요??** 
    - 전달용이므로 `controller`에 위치시키는게 맞지만 고민을 해도 답이 잘 안나옵니다!

2. 같은 도메인인 `Cars`와 `Names`는 전달할 일이 없으므로 **서로 의존하여 필요시 `getter`를 사용해도 될까요?**
    - SQL의 join처럼 괜찮을 것 같습니다만.. getter를 지양하라.. 라는 객체지향 관점에서.. 잘 모르겠습니다!
    ```java
    //Cars
    public static Cars fromNames(Names names) {
        return new Cars(names.getNames().stream()
                        .map(name -> Car.fromName(name))
                        .collect(Collectors.toList()));
    }
    
    
    
    
    //Names 
    public List<Name> getNames() {
        return this.names;
    }
    ```

3. `Car`는 이제 `Name`과 `Position`이라는 VO를 가지게 되었는데, Name과 Position도 Dto가 필요한지 궁금합니다.
    - CarDto List만 넘겨주면, OutputView에서는 cardto.getName() -> Name객체 -> .toString() -> string으로 원하는 값을 체인하여 Name, Position에 의존하지 않고 사용할 수 있습니다.
    - 하지만, Name객체속 값을 얻어낼 때도 Car처럼.. dto를 통해 가져와야할지 궁금합니다!

    ```java
    //OutputView.java
    
    public static void printRacingRecord(List<CarDto> cars) {
        for (CarDto car : cars) {
            System.out.println(car.getName().toString() + NAME_AND_SCORE_DELIMITER + printDash(car.getPosition().toInt()));
        }
        System.out.println();
    }
    ```


**아직 온보딩 기간이라 다른 것들과 병행하며 공부를 하고 있습니다!** 리뷰 받는 동안 추가적으로 적용해보겠습니다!

wishoon committed:
feat(docs): 요구사항에 따른 기능구현 목록 작성

wishoon committed:
test: 문자열 분리 테스트

wishoon committed:
test: 문자열 자르기 테스트

wishoon committed:
test: 문자의 특정 위치 예외 처리 테스트

wishoon committed:
test: 문자열 자르기 테스트 리팩토링

wishoon committed:
test: 집합 크기 확인 테스트

wishoon committed:
test: 집합 원소 포함 확인 테스트

wishoon committed:
test: 집합 원소 포함, 미포함 테스트

wishoon committed:
test: 빈 문자열 또는 null 입력시 0 반환 테스트

wishoon committed:
feat: 숫자 1개 변환 기능 구현

wishoon committed:
feat: 숫자 2개를 콤마로 구분하여 덧셈하는 기능 구현

wishoon committed:
feat: 덧셈기능에 콜론 구분자 추가 구현

wishoon committed:
feat: 커스텀 구분자를 지정하여 덧셈하는 기능 구현

wishoon committed:
feat: 음수를 전달할 경우 예외가 발생하는 기능 구현

wishoon committed:
feat: 기본 구분자와 커스텀 구분자 혼합 예외처리 기능 구현

Commit:
docs: 자동차경주 기능 구현 목록 작성

Commit:
Revert "docs: 자동차경주 기능 구현 목록 작성"

This reverts commit 155dfe5434e709e4de8e3fe884792af0e3c905a7.

Commit:
push test

is2js committed:
docs: 자동차경주 기능 구현 목록 작성

is2js committed:
Merge remote-tracking branch 'origin/step1' into step1

wishoon committed:
refactor: 문자열 덧셈 계산기 리팩토링

wishoon committed:
chore: 구현기능 목록 디렉토리 변경

wishoon committed:
feat: 각 자동차들의 이름을 입력 받는 기능 구현

wishoon committed:
feat: 자동차 경주 게임의 시도할 횟수를 입력 받는 기능 구현

wishoon committed:
docs: 구현 기능 목록 수정

wishoon committed:
feat: 랜덤한 값으로 전진 값을 생성하는 기능 구현

wishoon committed:
feat: 자동차 게임을 생성하는 기능 구현

wishoon committed:
feat: 자동차 게임의 시도 횟수 카운팅 기능 구현

wishoon committed:
feat: 자동차 게임의 실행 기능 구현

wishoon committed:
feat: 자동차 게임의 라운드별 기록을 가져오는 기능 구현

is2js committed:
Merge branch 'step1' of https://github.com/Wishoon/java-racingcar into step1

wishoon committed:
feat: 자동차 게임의 라운드별 기록을 출력하는 기능 구현

wishoon committed:
fix: 커스텀 구분자를 등록하지 못하는 에러 수정

wishoon committed:
refactor: 정적 팩토리 메서드 패턴에서 변환하도록 변경

wishoon committed:
feat: 가장 많은 전진을 한 자동차를 구하는 기능 구현

is2js committed:
Merge branch 'step1' of https://github.com/Wishoon/java-racingcar into step1

wishoon committed:
feat: 자동차 게임의 우승자 판별 기능 구현

wishoon committed:
feat: 자동차 게임의 우승자 판별 기능 구현

wishoon committed:
feat: 실행결과 문자열 콘솔 출력 기능 구현

wishoon committed:
docs: 구현 기능 목록 수정

wishoon committed:
feat: 자동차 이름 중복검사 기능 구현

wishoon committed:
refactor: 자동차 이름의 공백 검사 기능 변경

wishoon committed:
fix: " " 상수로 변경

wishoon committed:
refactor: 값 객체로 변경

wishoon committed:
chore: Util 페키지 디렉토리 변경

is2js committed:
Merge branch 'step1' of https://github.com/Wishoon/java-racingcar into step1

is2js committed:
chore: 불필요한 클래스(Main) 제거

is2js committed:
chore: 선언 순서대로 메소드 재배치

is2js committed:
refactor: 상수 네이밍 변경

is2js committed:
refactor: isContinue -> hasNextGame 메서드명 변경

is2js committed:
test: 테스트 메서드 DisplayName 명시적으로 변경

is2js committed:
refactor: 특정 상수의 접근제어자 수정 및 Round의 메서드명 변경

is2js committed:
refactor: 메서드 네이밍 변경(appendComma -> collectingResult)

is2js committed:
refactor: InputView내 검증 메소드들을 Util화

is2js committed:
feat: Car 생성자에 검증 로직 추가함

is2js committed:
refactor: print메서드 InputView->OutputView로 이동

is2js committed:
style: 축약된 네이밍 수정(o->object)

is2js committed:
refactor: round 생성자 매개변수 type 변경

is2js committed:
refactor: 불필요한 생성자 제거함

is2js committed:
refactor: inputView 내부 print문을 OutputView메서드로 추출

is2js committed:
test: Car, Round 생성자를 통한 검증 테스트 추가

is2js committed:
style: CODE STYLE(우테코-intellij-xml) 및 CheckStyle(네이버핵데이-java-xml) 적용

is2js committed:
docs: readme.MD 위치 변경(root->root/docs)

is2js committed:
refactor: StringCalculator 패키지명 변경

is2js committed:
test: 생성자 검증 추가에 따른 테스트 코드 수정

is2js committed:
refactor: 불필요한 생성자 삭제에 따른 리팩토링 및 일부 TEST 수정

is2js committed:
refactor: 추상화된 클래스명 구체화(Utils -> Validators)

is2js committed:
refactor: 관련된 Validators들을 묶어서 호출하는 메서드 생성

is2js committed:
style: 불필요한 공백 제거

is2js committed:
style: Car 정적 팩토리 메서드 네이밍 변경(from->fromNames)

is2js committed:
style: 클래스명 변경(Validators->Validator)

is2js committed:
feat: Cars 생성자 추가 생성 및 그에 따른 Test 수정

is2js committed:
feat(Round): 불필요한 생성자 제거 및 그에 따른 Test코드 수정

is2js committed:
fix: Round 생성자 수정을 통한 프로그램 예외처리 해결

is2js committed:
Merge branch 'step1' into step2

# Conflicts:
#	src/main/java/racingcar/domain/RacingGame.java
#	src/main/java/racingcar/domain/Round.java
#	src/main/java/racingcar/util/Validator.java
#	src/main/java/racingcar/view/InputView.java
#	src/test/java/racingcar/domain/RoundTest.java

is2js committed:
refactor: 불명확한 객체 전달 구조 제거후 DTO class 추가

is2js committed:
refactor: CarDto 도입으로 Cars 기능 및 Test코드 수정

is2js committed:
refactor: RacingResult 삭제로 인한 구조 리팩토링

is2js committed:
feat: 테스트를 위한 고정 숫자 생성 전략객체 생성 후 Test 적용

is2js committed:
refactor: Round View<->Model 검증 로직을 나누기 위한 구조 변경

is2js committed:
style: validator 메소드명 변경

is2js committed:
refactor: Cars View<->Model 검증 로직을 나누기 위한 구조 변경

is2js committed:
refactor: Car View<->Model 검증 로직을 나누기 위한 구조 변경

is2js committed:
feat: 시도 횟수 양수 검증은 View에서만 이루어지도록 메서드 추가

is2js committed:
refactor: main메서드 RacingController -> Application으로 이동

is2js committed:
feat: 원시값 name, carNames를 포장한 VO, 일급컬렉션 생성

is2js committed:
fix: Name, Names 생성자 private으로 수저ㅜㅇ

is2js committed:
test: Name class에 대한 Test코드 추가

is2js committed:
feat(Names): View와 중복 검증을 피하기 위한 생성자 추가

is2js committed:
test(NamesTest): Names 테스트 코드 추가

is2js committed:
feat: Position(VO) 생성하여 원시값 포장

is2js committed:
test: Position에 대한 테스트 코드 추가

is2js committed:
refactor: Position(VO) 도입에 따른 리팩토링

choihz commented on src/main/java/racingcar/util/Validator.java:
```diff
@@ -3,65 +3,51 @@
+        for (String carName : carNames) {
+            validateInput(carName);
         }
     }
```
1. `,,`처럼 입력하면 어떻게 되나요?
2. `a, a`처럼 공백이 들어가면 어떻게 되나요?

is2js replied:
1. `,,`의 검증은 split이후 개별 carName별로 validateInput 걸려서 문제가 없을까 생각했었는데
    1. split()에 2번째 인자(-1)를 주지 않으면 `,,`가 모두 다 쪼개지질 않는다 -> **split( , -1)으로 수정**
    ```java
        String[] split = ",,".split(","); // count 0
        String[] split1 = ",,a".split(","); // 3
        String[] split2 = ",a,".split(","); // 2
        String[] split3 = "a,,".split(","); // 1
    ```

2. split(,-1)이후 trim()하여, 구분자 주위의 공백을 허용( trim후 개별 carName으로 전달)하도록 수정하였습니다.
    ```java
      public static List<String> splitWithDelimiter(String value, String delimiter) {
          return Arrays.stream(value.split(delimiter, -1))
                  .map(String::trim)
                  .collect(Collectors.toList());
    }
    ```

감사합니다 코니!

choihz commented on src/main/java/racingcar/util/Validator.java:
```diff
@@ -3,65 +3,51 @@
-        if (validateCarsNameSize(carsName)) {
-            throw new IllegalArgumentException(ERROR_CARS_NAME_DUPLICATED);
+    private static void checkDuplicated(List<String> carNames) {
+        if (carNames.stream().distinct().count() != carNames.size()) {
```
`carNames.stream().distinct().count()` 같은 코드는 단번에 의미를 파악하기 어렵죠. 변수로 분리해 이름을 부여해 준다면 이해하기 더 수월할 것 같아요.

is2js replied:
~문 전체가 아니더라도 이렇게 연쇄적인 코드의 변수는 추출하여 이름을 부여하면 더 좋군요! 감사합니다 코니! 수정하였습니다!

`long distinctCountOfCarNames = carNames.stream().distinct().count();`

choihz commented on src/main/java/racingcar/util/Validator.java:
```diff
@@ -3,65 +3,51 @@
+    private static final String ERROR_NULL_EMPTY_MESSAGE = "빈칸 입력은 허용하지 않는다.";
+    private static final String ERROR_INCLUDE_BLANK_MESSAGE = "내부에 공백이 포함될 수 없습니다.";
+    private static final String ERROR_DUPLICATED_MESSAGE = "중복값을 입력할 수 없습니다.";
+    public static final String ERROR_INVALID_FORMAT_MESSAGE = "입력한 값이 숫자의 형태가 아닙니다.";
```
접근 제어자 확인해 주세요~

is2js replied:
코니, intellij 에서 상수 추출시 기본적으로 `public`으로 뽑아내는 것을 -> `private`으로  추출하도록 수정하였습니다! 

감사합니다 코니!

choihz commented on src/main/java/racingcar/util/Validator.java:
```diff
@@ -3,65 +3,51 @@
-    public static void validateCarName(String carName) {
-        validateCarsNameLength(carName);
-        validateCarsNameBlank(carName);
+    public static void validateInput(String carName) {
```
여기서만 사용되는 메서드 같은데 `public`으로 열어둘 필요가 있을까요?

is2js replied:
네 코니! 
validateInput()과 validateInputNumber()는 이후 확장성 모든 콘솔input들이 거쳐야할 공통 검증 로직으로 설계하여 열어두었는데

한 곳에서 사용되고 끝난 것에 대해서는 `private`로 수정하겠습니다

choihz commented on src/main/java/racingcar/util/Validator.java:
```diff
@@ -3,65 +3,51 @@
 import java.util.List;
 
 public class Validator {
```
이처럼 `static` 멤버만 가지고 있어 객체를 생성하여 사용할 필요가 없는 클래스의 경우, 명시적으로 `private` 기본 생성자를 선언해 불필요한 객체 생성을 막기도 한답니다.

is2js replied:
유틸성 클래스에서 생성해놓는 습관을 가져보도록 하겠습니다!
Validator 뿐만 아니라 Util클래스에서도 생성해두었습니다!

감사합니다! 코니!

choihz commented on src/main/java/racingcar/util/Validator.java:
```diff
@@ -3,65 +3,51 @@
+        if (inputRound.trim().contains(BLANK)) {
+            throw new IllegalArgumentException(ERROR_INCLUDE_BLANK_MESSAGE);
         }
     }
```
이 메서드는 `inputRound`를 검증하는 데만 사용되는 게 아닌데 파라미터 이름을 다시 생각해야 하지 않을까요?

is2js replied:
input에 대한 공통검증 메서드들에 대해서 작성시 인자들의 명칭을 메서드입장에서 고려해서 짓는 버릇을 들이겠습니다

감사합니다 input과  inputNumber(문자열로 들어오는 숫자)으로 다 수정하겠습니다!

choihz commented on src/main/java/racingcar/util/Validator.java:
```diff
@@ -3,65 +3,51 @@
+        checkNullOrEmpty(inputRound);
+        checkIncludeBlank(inputRound);
+        checkValidFormat(inputRound);
     }
```
여기서 `checkIncludeBlank(inputRound);` 작업이 꼭 필요한가요? 해당 줄을 주석 처리하고 횟수에 공백을 넣어 볼까요?

is2js replied:
`1 1`등의 입력을 방지하기 위해 사용했는데, format검사를 넣는 경우 필요 없다는 것을 깨달았습니다!

문자열 입력받을 땐 필요하나, 숫자입력시 포맷검사한다면 검증할 필요가 없군요!

감사합니다 코니!

choihz commented on src/main/java/racingcar/util/Util.java:
```diff
@@ -5,11 +5,13 @@
+
+    public static int convertToInt(String inputRound) {
+        return Integer.parseInt(inputRound);
+    }
```
유용하게 사용하려고 따로 분리하신 것이겠지만, 만약 코드의 다른 부분에서 `Integer.parseInt()` 작업이 필요하다면 굳이 이 메서드를 찾아서 쓸까요?

is2js replied:
api가 잘 제공될 때는 메서드를 따로 정의안하여 쓸데없는 지식을 피하겠습니다!

choihz commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -33,15 +32,15 @@ private static String printDash(int position) {
-        return racingResult.getRacingRecord().stream().map(Car::getName).collect(Collectors.joining(NAME_DELIMITER));
+    private static String collectingResult(List<CarDto> cars) {
+        return cars.stream().map(car -> car.getName().toString()).collect(Collectors.joining(NAME_DELIMITER));
     }
```
스트림 사용 시 이렇게 한 줄로 길게 늘여쓰면 가독성이 좋지 않은 것 같은데, 돌범의 의견은 어떤지 궁금하네요.

is2js replied:
`.`은 한 줄당 1개의 체인만 보여야된다고 프리코스에서 학습했던 것 같습니다!

검증하는 stream의 경우 한 줄로 적었던 적이 있는데, 팀 컨밴션에 맞게, 우테코에서 권유하는 방식에 맞게 상기 컨밴션을 따르도록 수정하겠습니다!

choihz commented on src/main/java/racingcar/controller/CarDto.java:
```diff
@@ -0,0 +1,27 @@
+
+    public Position getPosition() {
+        return position;
+    }
```
DTO의 사용 목적을 다시 생각해 보고 이 구조에 대해서도 고민해 볼까요?
`Name`, `Position` 객체의 존재를 DTO를 받는 쪽에서 알아야만 할까요? 어차피 원시값을 꺼내 사용하고 있지 않나요?

choihz replied:
> `Car`는 이제 `Name`과 `Position`이라는 VO를 가지게 되었는데, Name과 Position도 Dto가 필요한지 궁금합니다.

VO와 DTO의 개념에 대해 다시 학습해 보고, 왜 이것들을 사용하는지 스스로도 고민해 보면 이 질문의 답을 직접 찾으실 수 있을 것 같아요.

choihz commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,40 +1,43 @@
 
 public class Car implements Comparable<Car> {
-    private static final int MINIMUM_DRIVE_CONDITION = 4;
+    private static final int MOVING_DEMANED_MINIMUM_NUMBER = 4;
```
오타가 있네요~

is2js replied:
밑에 알려주신 상수 순서 컨밴션과 함께, ide가 알려주는 오타 체크도 눈 크게 뜨고 꼼꼼하게 보도록 하겠습니다!

choihz commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,40 +1,43 @@
-    private final String name;
-    private int position = DEFAULT_POSITION_VALUE;
+    private final Name name;
+    private Position position;
```
상수와 멤버 변수는 의미가 많이 다르기 때문에, 그 사이에 개행이 있으면 코드 가독성이 좋아져요.

is2js replied:
코니! 하나부터 열까지 지적하도록 한 것 같아 죄송합니다 ㅠ_ㅠ
감사합니다 작은 것부터 기록해서 제 것으로 만들어가겠습니다

choihz commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,40 +1,43 @@
 
-    public Car(String name) {
+    private Car(Name name) {
         this.name = name;
-        Validator.validateCarName(this.name);
+        this.position = Position.from(DEFAULT_POSITION_VALUE);
+    }
+
+    public static Car fromName(Name name) {
+        return new Car(name);
     }
```
개인적으로, 여전히 이 정적 팩터리 메서드의 존재 의의는 의문이긴 하네요 ㅎㅎ 꼭 바꿔야 한다는 건 아니고, 돌범이 느끼신 특별한 장점이 있는지 궁금해요.

is2js replied:
.fromName()은 Name으로부터  Car를 생산하는 곳인데,  Name -> Car로 변환되기 전 검증로직 등 있으면 추가할 수 있습니다.

**원래 `.fromName`은  Name이라는 원시값 포장을 도입하기 전, `String name`에 대한 검증이 이루어졌던 곳입니다.** 

현재는 name에 대한 자체 검증을 Name 내부로 옮겼기 때문에, 검증은 하지 않지만, 
- **Name이라는 객체에서부터 생산된다는 의미를 내포**한다는 점
- 도메인 테스트를 위해 필요한 public한 생성자를 대신한다는 점(검증없이 100% 생산해주는 private 기본 생성자를 유지)
- 빠트린 Car에 대한 검증이 추가될 수도 있다는 점에서 아직 살려두었습니다.

choihz commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,40 +1,43 @@
-        return name;
+    public Name toName() {
+        return this.name;
     }
```
`toPosition()`, `toName()` 같은 이름은 전치사 to의 의미를 생각했을 때 적절하지 않아 보이네요. `toString()`과 비교해서 생각해 볼까요?

is2js replied:
먼저 toString()의 `to`는 자료형을 바꿀 때 사용하네요! 
네이밍에 의심이 간다면, 비슷하게 사용된 메서드들을 떠올려 보도록 하겠습니다.

위의 코드에서는 자료형 전환이 필요한 것이 아니므로 `getNameValue()`, `getPositionValue()`으로 수정하였습니다!
- CarDto에서 Name객체 Position객체를 가지고 오는 getName(), getPosition()이 이미 존재하고 연쇄되어 사용되기 때문에 위와 같이 지었습니다. 
 
혹시 다른 네이밍 방법이 있을까 궁금합니다~!

```java
public static void printRacingRecord(List<CarDto> cars) {
    for (CarDto car : cars) {
        System.out.println(
                car.getName().getNameValue() + NAME_AND_SCORE_DELIMITER + printDash(
                        car.getPosition().getPositionValue()));
    }
    System.out.println();
}
```

choihz commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -1,40 +1,43 @@
     public int compareTo(Car car) {
-        return this.position - car.position;
+        return this.position.toInt() - car.position.toInt();
     }
```
`Position`을 이용해 비교한다면, `Position` 자체에도 `compareTo()`가 있는 것이 자연스럽지 않을까요?

is2js replied:
Position도 comparable을 구현하여, 비교가능하도록 수정하였습니다!

choihz commented on src/main/java/racingcar/domain/FixedNumberGeneratePolicy.java:
```diff
@@ -0,0 +1,10 @@
+    public static final int FIXED_NUMBER = 4;
+
+    @Override
+    public int generateNumber() {
```
인터페이스의 이름이 이미 `NumberGeneratePolicy`이기 때문에, 메서드 이름은 `generate()` 정도로 해도 충분할 것 같아요.

is2js replied:
인터페이스, 클래스명(Type)과 메서드의 이름에 목적어 중복이 안되도록 해야겠네요 !

수정하겠습니다~ 감사합니다 코니!

choihz commented on src/main/java/racingcar/domain/Name.java:
```diff
@@ -0,0 +1,55 @@
+
+    private static void checkValidLengthOfName(String name) {
+        int nameLength = name.length();
+        if (!(CAR_NAME_MIN_LENGTH <= nameLength && nameLength <= CAR_NAME_MAX_LENGTH)) {
```
조건문을 더 간단하게 만들 수 있지 않을까요? 부정 조건문은 의미를 단번에 파악하기가 어려워요.

is2js replied:
저의 경우,  `이게 아닌 경우 -> 에러를 발생시켜!`가 더 편해서 이런 식으로 작성하였는데 ㅠ 부정조건문이라는 말 자체도 익숙하지가 않네요.

이번 기회에 수정하고 지양하는 방향으로 가보겠습니다!

choihz commented on src/main/java/racingcar/controller/CarDto.java:
```diff
@@ -0,0 +1,27 @@
+import racingcar.domain.Name;
+import racingcar.domain.Position;
+
+public class CarDto {
```
> CarDto도 `domain`패키지에 넣어야할까요??

아니요. 지금 위치가 더 적절한 것 같아요.

is2js replied:
네 Dto에 대해서 더 알아봤는데, 

자동차경주에 대해서는 controller가 사용하는 것
5 Layer 구조에서는 service가 사용하는 것이 더 낫다는 글을보았습니다!

choihz commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -13,29 +12,25 @@ private Cars(List<Car> cars) {
+        return new Cars(names.getNames().stream()
+                .map(Car::fromName)
                 .collect(Collectors.toList()));
     }
```
> 같은 도메인인 `Cars`와 `Names`는 전달할 일이 없으므로 서로 의존하여 필요시 `getter`를 사용해도 될까요?

필요할 땐 써야죠. 무조건 "getter를 쓰면 안 돼"라고 생각하실 게 아니라, 이런 이야기가 나오는 이유가 무엇인지를 깊이 생각해 보셨으면 좋겠습니다.

is2js replied:
금일 네오의 자동차경주 피드백에서 **도메인내 객체간 협력**시 메세지를 던지자. -> getter를 지양하자는 이야기였던 것 같습니다! 

지금의 경우는, 이미 Names에게 새로운 결과물을 만들도록 시키는 경우가 아니라, 생성시 필요한 재료들을 꺼내쓰는 생성자내부이므로 getter가 필요할 것 같습니다!

choihz commented on src/main/java/racingcar/domain/Name.java:
```diff
@@ -0,0 +1,55 @@
+    private static final int CAR_NAME_MIN_LENGTH = 1;
+    private static final int CAR_NAME_MAX_LENGTH = 5;
+    private static final String ERROR_INVALID_CAR_NAME_LENGTH_MESSAGE =
+            CAR_NAME_MIN_LENGTH + "~" + CAR_NAME_MAX_LENGTH + " 글자 범위 내에서 입력하세요.";
```
선언 순서를 다시 고민해 볼까요?

is2js replied:
상수 (static final) -> final (인스턴스변수) 순으로 배열하고 중간에 빈줄 1개를 삽입하여 구분하겠습니다!

choihz reviewed (changes requested):
안녕하세요, 돌범! 리뷰어 코니입니다.

2단계도 잘 구현해 주셨네요! 피드백을 보시고, 더 궁금한 점이나 의견이 있으시다면 언제든 코멘트 또는 DM 주세요~

is2js committed:
feat: 빈값이 포함된 입력이 검증되도록 관련 메서드 수정

- Util#splitWithDelimiter split()에 2번째 인자 -1 추가
- Validator#validateCarNames에 checkCarNamesEmpty()메서드 추가

is2js committed:
refactor: Validator 및 Util 피드백 반영한 리팩토링

- Validator 객체 생성 불필요 명시를 위한 private 생성자 추가
- checkDuplicated 변수 분리를 통한 의도 명확화
- validateInput 사용 범위에 맞게 접근제어자 변경(public->private)
- 인자 명칭이 지나치게 구체적인 것을 추상화

is2js committed:
refactor: 메서드명(generateNumber)에서 Type에 명시된 목적어 제거

is2js committed:
refactor: Position Comparable 구현에 따른 객체 비교 리팩토링 및 불명확한 네이밍 수정

is2js committed:
refactor: 과도한 메서드 추출 제거

choihz commented on src/main/java/racingcar/controller/RacingGameController.java:
```diff
@@ -1,23 +1,42 @@
+
+    private Cars getCars() {
+        try {
+            return Cars.fromNames(Names.fromInput(InputView.getCarNames()));
```
지금은 `Cars`를 만들기 위해 `Names`의 존재를 알아야 하는 상황인데요, 사실 `Names`는 `Cars` 안에서만 사용되고 있죠. 그렇다면 `Names`의 존재가 바깥에 노출될 필요가 있을까요? 구현을 드러내고 있는 건 아닐까요?

is2js replied:
같은 context에서 사용될 일이 없는, 객체 생성 가운데 활용되는 객체들은 구현을 드러내지 않는 것이 좋을 것 같습니다!

Cars의 생성자명을 fromNames - fromInput으로 수정하고
그 내부에서 Names를 만들고 -> Car -> CarList를 만드는 순으로 정리해보았습니다.  Names가 굳이 필요없을 것 같긴한데, 원시값 포장의 연습이라고 생각해야할지.. 흠.. 고민이 됩니다.

```java
public static Cars fromInput(List<String> InputNames) {
    Names names = Names.fromInput(InputNames);
    return new Cars(names.getNames().stream()
            .map(Car::fromName)
            .collect(Collectors.toList()));
}
```

choihz commented on src/main/java/racingcar/util/Util.java:
```diff
@@ -5,11 +5,13 @@
+        return Arrays.stream(value.split(delimiter, -1))
+                .map(String::trim)
                 .collect(Collectors.toList());
     }
```
> 유용하게 사용하려고 따로 분리하신 것이겠지만, 만약 코드의 다른 부분에서 `Integer.parseInt()` 작업이 필요하다면 굳이 이 메서드를 찾아서 쓸까요?

개인적으로는 이 메서드도 위 코멘트의 대상이었던 메서드와 유사한 성격으로 보입니다.

is2js replied:
convert 뿐만 아니라 split + trim정도는 불필요하게 메서드로 추출하지 않도록 수정하겠습니다~!

choihz commented on src/main/java/racingcar/util/Validator.java:
```diff
@@ -3,65 +3,62 @@
+        if (input.trim().contains(BLANK)) {
+            throw new IllegalArgumentException(ERROR_INCLUDE_BLANK_MESSAGE);
         }
     }
```
메서드를 사용 순서로 배치중이신 거라면 이 부분의 재정렬이 필요해 보이네요.

is2js replied:
메서드 순서대로 재배치하였습니다. 수정하고 나서 다시 체크하는 것이 버릇이 안들어있습니다.
코니만의 노하우가 있을까요??

intellij의 Structure 보는 기능을 활용해보려 합니다 감사합니다 코니!

choihz commented on src/main/java/racingcar/domain/Name.java:
```diff
@@ -0,0 +1,60 @@
+
+    private static void validateName(String name) {
+        checkValidLengthOfName(name);
+    }
```
이 메서드가 필요한가요?

is2js replied:
검증 메서드 확장을 고려했는데, 더이상 요구사항은 없으므로 삭제하겠습니다! 감사합니다~!

choihz commented on src/main/java/racingcar/domain/FixedNumberGeneratePolicy.java:
```diff
@@ -0,0 +1,10 @@
+package racingcar.domain;
+
+public class FixedNumberGeneratePolicy implements NumberGeneratePolicy {
```
이 클래스는 테스트 코드에서만 사용되고 있는데, 여기에 위치하는 것이 적절할까요? 또한, 자바의 람다식에 관해 학습해 보셔도 좋을 것 같아요.

is2js replied:
초기에는 람다식을 통해 () -> 5를 전략객체에 대입하였다가, 실제 다른 전략객체 확장을 해보고 싶어서 작성하였습니다.!

테스트에만 사용되는 전략객체는 테스트에 위치시키는게 좋을 것 같습니다! 해당 클래스는 삭제하고 람다식으로 대체하겠습니다!

choihz commented on src/main/java/racingcar/domain/Position.java:
```diff
@@ -0,0 +1,60 @@
+package racingcar.domain;
+
+public class Position implements Comparable<Position> {
+    private static final String ERROR_POSITION_INVALID_RANGE_MESSAGE = "정상 범위(" + 0 + "이상)가 아닙니다";
```
혹시 메시지를 이렇게 구성하신 이유가 무엇일까요?

is2js replied:
가운데 매직넘버를 미리 선언된 상수로 대체하려고 하였는데, ㅠ_ㅜ 둘다 놓친 것 같습니다.

매직넘버의 상수만 수정되면 자동으로 메세지도 수정되도록 고쳤습니다

감사합니다~!

choihz commented on src/main/java/racingcar/domain/Position.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private static void checkValidRangeOfPosition(int position) {
+        if (!(MIN_POSITION_VALUE <= position)) {
```
```suggestion
        if (MIN_POSITION_VALUE > position) {
```

위 코드와 동일하지 않나요?

> 저의 경우,  `이게 아닌 경우 -> 에러를 발생시켜!`가 더 편해서 이런 식으로 작성하였는데 ㅠ 부정조건문이라는 말 자체도 익숙하지가 않네요.

이 코멘트에 남겨주신 생각에 따라 이런 방식으로 작성하신 것 같기도 한데, 이 부분에선 보편적인 방향으로 생각을 수정하시는 편이 좋을 것 같습니다.

```suggestion
        if (position < MIN_POSITION_VALUE) {
```

그리고 순서도 이 편이 더 읽기 좋아요.

is2js replied:
부정조건문을 제거하고 상수 위치를 바꾸고, 매직넘버를 선언된 상수로 교체하였습니다!

choihz commented on src/main/java/racingcar/domain/RacingGame.java:
```diff
@@ -13,14 +14,14 @@ public RacingGame(Cars cars, Round round, NumberGeneratePolicy numberGeneratePol
     }
 
-    public static RacingGame createRacingGame(List<String> carsName, String round,
+    public static RacingGame createRacingGame(Cars cars, Round round,
```
여기서 굳이 정적 팩터리 메서드를 사용하실 거라면, 생성자의 접근 제어자를 바꿔야 할 것 같아요.

is2js replied:
기본 생성자를 private으로 변경하였습니다!

조금 더 복잡한 어플리케이션이었다면, cache를 여기다가 적용해도 좋을 것 같습니다~!
감사합니다 코니~!

choihz commented on src/main/java/racingcar/util/Validator.java:
```diff
@@ -3,65 +3,62 @@
-    private static void validateCarsNameLength(String carName) {
-        if (carName.length() > CAR_NAME_LIMIT) {
-            throw new IllegalArgumentException(ERROR_CAR_NAME_LENGTH);
+    private static void checkDuplicated(List<String> carNames) {
+        long distinctCountOfCarNames = carNames.stream().distinct().count();
+        if (distinctCountOfCarNames != carNames.size()) {
+            throw new IllegalArgumentException(ERROR_DUPLICATED_MESSAGE);
         }
     }
```
이러한 방식도 괜찮지만, 우리의 요구사항(중복을 체크하자, 또는 중복 여부를 체크하자)을 들어줄 수 있는 자료구조를 활용하는 방법도 있어요. 참고만 해주세요.

is2js replied:
네 Set으로 프리코스 때 구현해본 것 같습니다!

choihz commented on src/test/java/racingcar/view/OutputViewTest.java:
```diff
@@ -10,12 +10,14 @@
+    private Cars cars;
 
     @BeforeEach
     public void setUpStreams() {
```
테스트 메서드들의 네이밍 방식이 다른데, 하나로 통일하는 편이 좋지 않을까요?

is2js replied:
셋업 메서드들도 cameCase를 적용하였습니다.

ㅠ_ㅜ 페어 프로그래밍의 거의 모든 부분이 드러내졌네요 ㅎㅎ
**다음 미션때는  꼼꼼한 코니의 티칭이 잘 반영하여 페어 프로그래밍 시작부터 잘해보겠습니다!**

choihz commented on src/test/java/racingcar/controller/CarDtoTest.java:
```diff
@@ -0,0 +1,26 @@
+
+    @DisplayName("Position from CarDto 값 확인")
+    @Test
+    void get_valid_position_from_cardto() {
```
사소하고 별 것 아닌 것 같은 테스트가 예상치 못한 실수를 막아줄 때도 분명 있지만, 테스트 코드도 유지보수해야 할 코드라는 걸 염두에 두시는 것이 좋습니다. 즉, 의미 있는 테스트 코드를 작성하는 것이 좋습니다.

물론 "의미 있는"이라는 것이 굉장히 주관적이고, 또 장단점이 공존하기 때문에 상황에 맞게 적절히 판단해야 하겠죠 😊

is2js replied:
cardto의 구조를 바꿀 때 테스트도 유지보수 해야되는군요

dto에서 값을 확인하는 코드는 이미 유기적으로 간접테스트 되고 있으니 빼도록 하겠습니다!

choihz commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -18,9 +17,11 @@ public static void printRacingRecordHeadLine() {
+            System.out.println(
+                    car.getName().getNameValue() + NAME_AND_SCORE_DELIMITER + printDash(
+                            car.getPosition().getPositionValue()));
         }
```
getter를 연쇄적으로 사용하고 있는데, 이 방법밖에 없는지 고민할 필요가 있지 않을까요? [이 코멘트](https://github.com/woowacourse/java-racingcar/pull/381#discussion_r809244279)와 함께 생각해 볼까요?

> DTO의 사용 목적을 다시 생각해 보고 이 구조에 대해서도 고민해 볼까요?
> `Name`, `Position` 객체의 존재를 DTO를 받는 쪽에서 알아야만 할까요? 어차피 원시값을 꺼내 사용하고 있지 않나요?

is2js replied:
해당 부분에 대해 생각해보았습니다.
DTO는 값의 변형을 막으면서 전달하는 객체기 때문에, 중간의 wrapper들이 굳이 없어도 될 것 같다고 판단하였습니다!

dto에서는 wrapper클래스 사용없이 바로 값만 반환되도록 수정하겠습니다~!

이렇게 연쇄적으로 호출되는 것을 `디미터법칙 위배`  `열차전복사고` 등이라고 표현하는 것 같더라구요~! 지양하도록

CarDto의 생성자에서는 Car.getName-> Name.getNameValue() -> string 값 추출을 위해 연쇄과정이 필요할 것 같습니다. 여기선 사용해도 될까요??

```java
    public static CarDto from(Car car) {
        return new CarDto(car.getName().getNameValue(), car.getPosition().getPositionValue());
    }
```

choihz commented on src/main/java/racingcar/domain/Names.java:
```diff
@@ -0,0 +1,39 @@
+                .collect(Collectors.toList()));
+    }
+
+    public static Names from(List<Name> names) {
+        validateNames(names);
+        return new Names(names);
+    }
```
테스트 코드에서만 사용되고 있네요 🤔

is2js replied:
현재 inputView의 기본검증을 이미 끝난 `string`으로 입력되는 `fromInput`과 
테스트나 프로덕션 내부에서 생성되어,  기본검증이 필요한 단일객체`Name`으로 입력되는 `from`을 나누어서 구현하고 있습니다. **Cars**역시 동일한 로직으로 test에서만 생성되고 있습니다!

Name -> Names의 구현도 Car -> Cars처럼, 단일객체에서부터 생성될시 필수검증이 필요하다고 생각합니다!

inputView에서만 이루어지고 있는 검증과 동일하며, 콘솔입력 테스트를 대신할 수 있다는 장점도 가지고 있습니다!

choihz commented on src/test/java/racingcar/domain/NameTest.java:
```diff
@@ -0,0 +1,27 @@
+import org.junit.jupiter.params.provider.ValueSource;
+
+class NameTest {
+    @DisplayName("이름이 1자 미만, 5자 초과일 경우")
```
"이름이 1자 미만, 5자 초과일 경우" 어떻다는 것인지, 테스트 이름을 분명히 작성해 두면 좋습니다. 특히 테스트가 실패했을 때 원인을 더 빨리 파악할 수 있을 거예요.

is2js replied:
```java
@DisplayName("자동차 이름이 1자 미만, 5자 초과하여 유효하지 않은 경우")
```
형식으로 구체적으로 수정하였습니다.


**꼼꼼한 코니의 리뷰 덕분에 지나칠 수도 있는 것들에 대해 다시 돌아보게 되었습니다!**

choihz reviewed (changes requested):
안녕하세요, 돌범!

피드백을 잘 반영해 주셨네요 👍 조금 더 나아질 수 있는 부분과 생각해 보면 좋을 내용들이 있어 코멘트를 드렸어요. 이외에도 더 궁금한 점이나 의견이 있으시다면 언제든 코멘트 또는 DM 주세요~

is2js committed:
refactor: 과도한 메서드 추출 제거

is2js committed:
refactor(RacingGame): 불필요한 구현 노출 코드의 캡슐화

is2js committed:
refactor: Name 불필요한 래핑 메소드 삭제 및 Validator 메소드 순서 정렬

is2js committed:
refactor: 불필요한 전략패턴 구상클래스 삭제 및 람다식으로 대체

is2js committed:
refactor(Position): 상수 순서 재별 및 매직넘버 상수화, 부정조건문 제거

is2js committed:
refactor(RacingGame): 생성자의 접근제어자 변경

is2js committed:
test: 불필요한 테스트 제거, 일부 메서드 네이밍 수정

is2js committed:
refactor: CarDto 값만 상태값으로 가지도록 수정함.

is2js committed:
test(NameTest): 불명확한 DisplayName 수정함

is2js committed:
refactor(Round): 부정조건문을 일반조건문으로 수정, 매직넘버 상수화

choihz commented on src/main/java/racingcar/domain/Name.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    public static Name from(String name) {
+        checkValidLengthOfName(name);
```
이 작업은 여기서 하는게 적절할까요, 아니면 생성자에서 하는 것이 적절할까요?

choihz commented on src/main/java/racingcar/controller/CarDto.java:
```diff
@@ -0,0 +1,25 @@
+        this.position = position;
+    }
+
+    public static CarDto from(Car car) {
+        return new CarDto(car.getName().getNameValue(), car.getPosition().getPositionValue());
+    }
```
> CarDto의 생성자에서는 Car.getName-> Name.getNameValue() -> string 값 추출을 위해 연쇄과정이 필요할 것 같습니다. 여기선 사용해도 될까요??

`Car`의 `getName()`, `getPosition()`이 꼭 `Name`, `Position` 객체를 리턴하는 게 좋은지 고민해 보시면 어떨까요?

choihz commented on src/main/java/racingcar/domain/Names.java:
```diff
@@ -0,0 +1,39 @@
+        this.names = names;
+    }
+
+    public static Names fromInput(List<String> names) {
+        return new Names(names.stream()
+                .map(Name::from)
+                .collect(Collectors.toList()));
+    }
```
> 현재 inputView의 기본검증을 이미 끝난 `string`으로 입력되는 `fromInput`과 
테스트나 프로덕션 내부에서 생성되어,  기본검증이 필요한 단일객체`Name`으로 입력되는 `from`을 나누어서 구현하고 있습니다. **Cars**역시 동일한 로직으로 test에서만 생성되고 있습니다!

하지만 `fromInput()`에 검증이 끝난 값만 들어온다고 **가정**하는 것이 적절할까요?

choihz commented on src/test/java/racingcar/domain/NameTest.java:
```diff
@@ -0,0 +1,27 @@
+import org.junit.jupiter.params.provider.ValueSource;
+
+class NameTest {
+    @DisplayName("자동차 이름이 1자 미만, 5자 초과하여 유효하지 않은 경우")
```
```suggestion
    @DisplayName("자동차 이름이 1자 미만, 5자 초과하여 유효하지 않은 경우 예외가 발생한다")
```

이정도로 결과까지 명확하게 서술하면 더 좋습니다.

choihz commented on src/test/java/racingcar/domain/NameTest.java:
```diff
@@ -0,0 +1,27 @@
+class NameTest {
+    @DisplayName("자동차 이름이 1자 미만, 5자 초과하여 유효하지 않은 경우")
+    @ParameterizedTest
+    @EmptySource
+    @ValueSource(strings = {"", "Wishoon", "123456"})
```
`@ValueSource`의 빈 문자열과 `@EmptySource`가 같은 값을 제공하지 않나요?

choihz commented on src/test/java/racingcar/domain/NameTest.java:
```diff
@@ -0,0 +1,27 @@
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @DisplayName("자동차 이름이 1~5자로 유요한 경우")
```
오타가 있네요~

choihz reviewed (approved):
안녕하세요, 돌범!

피드백을 잘 반영해 주셔서 이번 미션은 이제 마무리하려고 합니다 👍 우테코의 첫 미션을 함께하게 되어 즐거웠고, 정말 고생 많으셨어요. 그럼, 다음에 다시 만나요~ 👋
